### DISCUSSION TITLE
How can I replace the service to it mock in DIC in PHPUnit WebTestCase?

### DISCUSSION THREAD

**Discussion by voskobovich** [symfony/symfony#46952] [Q&A]
Hi all ✋ I am writing a functional test in my Symfony app. I need to replace the service to mock in the DI container.

I found the [post](https://dev.to/nikolastojilj12/symfony-5-mocking-private-autowired-services-in-controller-functional-tests-24j4) that described how to do this for Symfony 5, but it doesn`t work for me. I got an error:

`The "App\Service\Conversation" service is already initialized, you cannot replace it.`

I found the [issue](https://github.com/symfony/symfony/issues/28063) with a similar problem, but this answer doesn`t help me 😔

I use Symfony 6.1 with PHP Unit 9.5. Can I replace the service in DIC in WebTestCase with these versions?
Who has any ideas how I can solve my task?

```
# bundles/MyBundle/config/services.php

use App\Service\Conversation;
use App\Service\ConversationInterface;

return static function (ContainerConfigurator $container) {
    $services = $container->services();
    $services->defaults()
        ->autoconfigure()
        ->autowire()
    ;

    $services->set(Conversation::class);
    $services->alias(ConversationInterface::class, Conversation::class);

    // ...
```

```
# config/services_test.yaml
services:
  _defaults:
   autowire: true
   autoconfigure: true

  App\Service\Conversation:
   public: true
```

```
# tests/Functional/ConversationControllerTest.php
// ...
public function testIndex(): void
{
   $client = static::createClient();
   $container = self::getContainer();

   $class = \App\Service\Conversation::class;

   /** @var ConversationInterface|MockObject $mock */
   $mock = $this->getMockBuilder($class)
     ->setConstructorArgs($this->extractConstructorParams($container, $class))
     ->onlyMethods(['say'])
     ->getMock()
   ;

   $container->set($class, $mock);
}
// ...
```

**Comment by MatTheCat** [comment]
I don’t really know what you are trying to do with the mock in your example but it may be simpler to create one as a concrete class and make it the `App\Service\Conversation` service in your `config/services_test.yaml` :thinking:
